### PR TITLE
[Autotune] Add SearchSpace Code

### PR DESCRIPTION
**Pull request by zhhsplendid**
This PR added some auto generate rule codes.

The main change includes:
1. Base class/interface of `auto_gen_rule`
2. MultiLevelTile auto gen rule as the first derived class.
3. Code in search_space.cc which demonstrates how to use the auto_gen_rule

Please ignore the search_space.cc, it is still under developing and not tested yet. Reviewer can review it in the near future. Some codes are in https://github.com/PaddlePaddle/CINN/pull/731

### CODE REVIEW

File: cinn/auto_schedule/search_space/auto_gen_rule/multi_level_tiling.h [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+
+  // Returns true if sche_block_realize is applicable by MultiLevelTiling
+  bool MeetCondition(const ir::ScheduleBlockRealize& sche_block_realize) const;
+
+  // Set ScheduleBlock's read_buffers and write_buffers
+  void AnalyzeScheduleBlockReadWriteBuffer(ir::ScheduleBlock* sche_block) const;
+
+  // Sample pair of integer type (a, b) such as a * b = extent
+  template <typename T>
+  std::pair<T, T> SampleTileSplit(T extent) const {
```
Comment (CtfGo) [inline cinn/auto_schedule/search_space/auto_gen_rule/multi_level_tiling.h:57]: 这几个函数接口需要为public吗

File: cinn/ir/ir_printer.cc [hunk 2]
```diff
@@ -407,9 +407,20 @@ void IrPrinter::Visit(const _BufferRange_ *x) {
   for (int i = 0; i < x->ranges.size(); i++) {
     if (i) os() << ", ";
     auto &range = x->ranges[i];
-    CHECK(range->lower_bound.defined());
-    CHECK(range->upper_bound.defined());
-    os() << range->lower_bound << ":" << range->upper_bound;
+    // CHECK(range->lower_bound.defined());
+    // CHECK(range->upper_bound.defined());
+    // os() << range->lower_bound << ":" << range->upper_bound;
```
Comment (CtfGo) [inline cinn/ir/ir_printer.cc:412]: 注释掉这部分与当前兼容吗

File: cinn/auto_schedule/search_space/auto_gen_rule/multi_level_tiling.h (hunk 1 above)
Comment (zhhsplendid) [reply cinn/auto_schedule/search_space/auto_gen_rule/multi_level_tiling.h:57]: 我觉得暂时就这里用到，而且在SearchSpace中对这些AutoGenRule的子类调用不通过这几个接口。我感觉将来如果有需要，迁移到工具文件，不用public？

File: cinn/ir/ir_printer.cc (hunk 2 above)
Comment (zhhsplendid) [reply cinn/ir/ir_printer.cc:412]: 与当前兼容，这里只是做打印用。不过这里注释而不是删除代码，是我的错。我整理代码时忘了删除。这个PR CI已经过很多了，后面也有依赖PR，我将在后面提一个单独PR修改。

**Review by CtfGo — APPROVED** [review]

**Pull request merged**